dolphin-dnn/src/main/java/edu/snu/dolphin/dnn/layerparam/provider/GroupCommParameterProvider.java

```diff
@@ -67,23 +56,17 @@ private GroupCommParameterProvider(
   }
 
   @Override
-  public synchronized void push(final LayerParameter[] parameterGradients) {
-    // do not store the input if it is not valid
-    if (!(parameterGradients == null || parameterGradients.length == 0)) {
-      parameterGradientList.add(parameterGradients);
-    }
-
-    if (++pushCount >= batchSize) {
-      pushCount = 0;
-
-      try {
-        parameterGradientReduceSender.send(parameterGradientList);
-      } catch (final NetworkException e) {
-        throw new RuntimeException("NetworkException while trying to send reduce", e);
-      } catch (final InterruptedException e) {
-        throw new RuntimeException("InterruptedException while trying to send reduce", e);
+  public synchronized void push(final int batchSize, final LayerParameter[] parameterGradients) {
+    try {
+      if (batchSize == 0 || parameterGradients == null || parameterGradients.length == 0) {
+        parameterGradientReduceSender.send(Collections.<Pair<Integer, LayerParameter[]>>emptyList());
+      } else {
+        parameterGradientReduceSender.send(Collections.singletonList(new Pair<>(batchSize, parameterGradients)));
```

Comment:
In my opinion, the performance boost that can be gained by using `Pair`s and applying summation at reduce functions worths more than the code clearness we can achieve by using `List<Pair>`s. I think `<0, LayerParameter[0]>` is best for an empty message, because we can just ignore the length of the layer parameter array entirely if the batch size is zero.